File: javascript/chapter_1_ArraysAndStrings/q1_4_palinperm.js
'use strict';

/*
* Problem: Palindrome Permutation
* Algorithm; Assumption: 0 < # of characters < # of characters < 257
*
*/


/*
*
* Solution 1 : No data structure used.
* Time:
*
*/

function palinPerm(a,b){
  if (a.length != b.length) {
    console.log("Not PalinPerm");
  }

  let x = a.toLowerCase().split("").sort();
  let y = b.toLowerCase().split("").sort();
  let t = 0;

  for (let i = 0; i < a.length; i++){
    if (x[i] == y[t]){
      t++;
    }
  else {
    console.log("False, Not a PalinPerm");
    break;
  }
  }
  console.log("True, PalinPerm:", a+",", b);
}

palinPerm("Tact Coa", "taco cat"); //True
palinPerm("Tact Coa", "atco cta"); //True

File: java/chapter_1_arraysAndStrings/q1_4_palinperm.java
// q1_4_Palindrome Permutation: variation of a palindrome

// Questions to ask Interviewer
// 1. 

// Assumption
// 1. Palindrome is a string that is the same forward and backwards
// 2. Permutation can be a variation
// 3. We need to have a even and odd count of characters


// Time complexity:
// O(n) where n is the length of the string

// Space complexity:
//

public class q1_4_palinperm {
	public static boolean isPermutationOfPalindrome(String phrase){
		int[] table = buildCharFrequencyTable(phrase);
		return checkMaxOneOdd(table);
	}

	/* Check that no more than one character has an odd count. */
	public static boolean checkMaxOneOdd(int[] table){
		boolean foundOdd = false;
		for (int count: table){
			if(count % 2 == 1){
				if (foundOdd){
					return false;
				}
				foundOdd = true;
			}
		}
		return true;
	}

	/* Map each character to a number. a -> 0, b -> 1, c -> 2, etc.
	 * This is case insensitive. Non-letter characters map to -1. */
	public static int getCharNumber(Character c){
		int a = Character.getNumericValue('a');
		int z = Character.getNumericValue('z');
		int val = Character.getNumericValue(c);
		if (a <= val && val <= z){
			return val - a;
		}
		return -1;
	}

	/* Count how many times each character appears. */
	public static int[] buildCharFrequencyTable(String phrase) {
		int[] table = new int[Character.getNumericValue('z') -
					Character.getNumericValue('a') + 1];
		for (char c : phrase.toCharArray()) {
			int x = getCharNumber(c);

			if (x != -1){
				table[x]++;
			}
		}
		return table;
	}

	// Insert Print Function Here
	// "taco cat"  --->  "atco cta"

}


